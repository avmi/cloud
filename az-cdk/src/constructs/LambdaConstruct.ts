import { Construct, Duration } from '@aws-cdk/core';
import { Code, Function, Runtime, ILayerVersion, LayerVersion } from '@aws-cdk/aws-lambda';
import { PolicyStatement } from '@aws-cdk/aws-iam';
import { Iany } from '@cpmech/basic';

export interface ILambdaProps {
  filenameKey: string; // example 'index' (without .js extension)
  handlerName: string; // example 'handler'
  envars?: Iany; // environmental variables passed to lambda function
  timeout?: Duration; // timeout
  memorySize?: number; // memory size
  runtime?: Runtime; // default = NODEJS_12_X
  dirDist?: string; // default = 'dist'
  layerArns?: string[]; // ARNs of needed layers
  extraPermissions?: string[]; // e.g. ['ses:SendEmail']
  accessDynamoTables?: string[]; // names of tables to give access
  accessBuckets?: string[]; // name of buckets to give access
}

export class LambdaConstruct extends Construct {
  readonly functionName: string;
  readonly functionArn: string;

  constructor(scope: Construct, id: string, props: ILambdaProps) {
    super(scope, id);

    let layers: ILayerVersion[] | undefined;
    if (props.layerArns) {
      layers = props.layerArns.map((arn) => LayerVersion.fromLayerVersionArn(this, 'Layer', arn));
    }

    const lambda = new Function(this, 'Lambda', {
      runtime: props.runtime || Runtime.NODEJS_12_X,
      code: Code.fromAsset(props.dirDist || 'dist'),
      handler: `${props.filenameKey}.${props.handlerName}`,
      layers,
      environment: props.envars,
      timeout: props.timeout,
      memorySize: props.memorySize,
    });

    this.functionName = lambda.functionName;
    this.functionArn = lambda.functionArn;

    if (props.extraPermissions) {
      props.extraPermissions.forEach((s) => {
        lambda.role?.addToPrincipalPolicy(
          new PolicyStatement({
            actions: [s],
            resources: ['*'],
          }),
        );
      });
    }

    if (props.accessDynamoTables) {
      const region = lambda.stack.region;
      props.accessDynamoTables.forEach((t) => {
        lambda.role?.addToPrincipalPolicy(
          new PolicyStatement({
            actions: ['dynamodb:*'],
            resources: [
              `arn:aws:dynamodb:${region}:*:table/${t}`,
              `arn:aws:dynamodb:${region}:*:table/${t}/index/*`,
            ],
          }),
        );
      });
    }

    if (props.accessBuckets) {
      props.accessBuckets.forEach((b) => {
        lambda.role?.addToPrincipalPolicy(
          new PolicyStatement({
            actions: ['s3:*'],
            resources: [`arn:aws:s3:::${b}`, `arn:aws:s3:::${b}/*`],
          }),
        );
      });
    }
  }
}
