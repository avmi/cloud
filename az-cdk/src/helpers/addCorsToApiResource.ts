import { IResource, MockIntegration, PassthroughBehavior } from '@aws-cdk/aws-apigateway';

export const addCorsToApiResource = (
  resource: IResource,
  methods: string[] = ['GET', 'PUT', 'POST', 'DELETE'],
) => {
  const m = methods.join(',');
  resource.addMethod(
    'OPTIONS',
    new MockIntegration({
      integrationResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Headers': `'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'`,
            'method.response.header.Access-Control-Allow-Origin': `'*'`,
            'method.response.header.Access-Control-Allow-Credentials': `'false'`,
            'method.response.header.Access-Control-Allow-Methods': `'OPTIONS,${m}'`,
          },
        },
      ],
      passthroughBehavior: PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': '{"statusCode": 200}',
      },
    }),
    {
      methodResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Headers': true,
            'method.response.header.Access-Control-Allow-Methods': true,
            'method.response.header.Access-Control-Allow-Credentials': true,
            'method.response.header.Access-Control-Allow-Origin': true,
          },
        },
      ],
    },
  );
};
