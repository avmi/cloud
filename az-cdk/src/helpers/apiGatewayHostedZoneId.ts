/*
  The API Gateway HostedZoneId for each region has to be looked up here:
     https://docs.aws.amazon.com/general/latest/gr/rande.html#apigateway_region

  Examples:
    US East (N. Virginia)      us-east-1       execute-api.us-east-1.amazonaws.com       Z1UJRXOUMOOFQ8
    US West (N. California)    us-west-1       execute-api.us-west-1.amazonaws.com       Z2MUQ32089INYE
    US East (Ohio)             us-east-2       execute-api.us-east-2.amazonaws.com       ZOJJZC49E0EPZ
    US West (Oregon)           us-west-2       execute-api.us-west-2.amazonaws.com       Z2OJLYMUO9EFXC
    Asia Pacific (Hong Kong)   ap-east-1       execute-api.ap-east-1.amazonaws.com       Z3FD1VL90ND7K5
    Asia Pacific (Singapore)   ap-southeast-1  execute-api.ap-southeast-1.amazonaws.com  ZL327KTPIQFUL
    Asia Pacific (Sydney)      ap-southeast-2  execute-api.ap-southeast-2.amazonaws.com  Z2RPCDW04V8134
    South America (Sao Paulo)  sa-east-1       execute-api.sa-east-1.amazonaws.com       ZCMLWB8V5SYIT
*/

export const apiGatewayHostedZoneId = (region: string) => {
  switch (region.toLowerCase()) {
    case 'us-east-2':
      return 'ZOJJZC49E0EPZ';
    case 'us-east-1':
      return 'Z1UJRXOUMOOFQ8';
    case 'us-west-1':
      return 'Z2MUQ32089INYE';
    case 'us-west-2':
      return 'Z2OJLYMUO9EFXC';
    case 'ap-east-1':
      return 'Z3FD1VL90ND7K5';
    case 'ap-south-1':
      return 'Z3VO1THU9YC4UR';
    case 'ap-northeast-2':
      return 'Z20JF4UZKIW1U8';
    case 'ap-southeast-1':
      return 'ZL327KTPIQFUL';
    case 'ap-southeast-2':
      return 'Z2RPCDW04V8134';
    case 'ap-northeast-1':
      return 'Z1YSHQZHG15GKL';
    case 'ca-central-1':
      return 'Z19DQILCV0OWEC';
    case 'eu-central-1':
      return 'Z1U9ULNL0V5AJ3';
    case 'eu-west-1':
      return 'ZLY8HYME6SFDD';
    case 'eu-west-2':
      return 'ZJ5UAJN8Y3Z2Q';
    case 'eu-west-3':
      return 'Z3KY65QIEKYHQQ';
    case 'eu-north-1':
      return 'Z3UWIKFBOOGXPP';
    case 'me-south-1':
      return 'Z20ZBPC0SS8806';
    case 'sa-east-1':
      return 'ZCMLWB8V5SYIT';
  }
  throw new Error(`cannot get API Gateway HostedId for region "${region}"`);
};
