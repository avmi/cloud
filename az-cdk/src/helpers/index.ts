export * from './apiGatewayHostedZoneId';
export * from './envars2cdk';
export * from './npmCommands';
export * from './ssmSecret';
export * from './synthApp';
