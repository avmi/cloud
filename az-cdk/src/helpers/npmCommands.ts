export interface INPMCommands {
  install: string[];
  build: string[];
}

export const npmCommands = (extraBeforeTest = '', useYarn = false): INPMCommands => {
  if (useYarn) {
    return {
      install: [
        'curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -',
        'echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list',
        'sudo apt-get update && sudo apt-get install yarn',
        'yarn',
      ],
      build: ['yarn test', 'yarn build'],
    };
  }
  return {
    install: ['npm install'],
    build: extraBeforeTest
      ? [extraBeforeTest, 'npm run test', 'npm run build']
      : ['npm run test', 'npm run build'],
  };
};
