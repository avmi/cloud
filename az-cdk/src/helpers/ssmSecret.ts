import { SecretValue, CfnDynamicReference, CfnDynamicReferenceService } from '@aws-cdk/core';

// ssmSecret creates a reference to a parameter in the SSM Parameter Store (not Secrets Manager!)
export const ssmSecret = (input: { name: string; version: string }): SecretValue => {
  return SecretValue.cfnDynamicReference(
    new CfnDynamicReference(CfnDynamicReferenceService.SSM, `${input.name}:${input.version}`),
  );
};
