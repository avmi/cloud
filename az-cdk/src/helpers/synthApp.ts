import { App } from '@aws-cdk/core';

// synthApp synthetizes App and returns object
export const synthApp = (app: App): any => {
  const assembly = app.synth();
  const synthesized = assembly.stacks[0];
  return synthesized.template;
};

// synthAppString synthetizes App and returns stringifyied object
export const synthAppString = (app: App): string => {
  return JSON.stringify(synthApp(app), undefined, 2);
};
