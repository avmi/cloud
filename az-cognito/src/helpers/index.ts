export * from './flattenAttributes';
export * from './email2username';
