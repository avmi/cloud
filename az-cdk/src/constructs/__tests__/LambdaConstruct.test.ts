import { App, Stack } from '@aws-cdk/core';
import { LambdaConstruct } from '../LambdaConstruct';
import { synthAppString } from '../../helpers/synthApp';

describe('LambdaConstruct', () => {
  it('synthetizes properly', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new LambdaConstruct(stack, 'Lambda', {
      filenameKey: 'index',
      handlerName: 'handler',
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });
});
