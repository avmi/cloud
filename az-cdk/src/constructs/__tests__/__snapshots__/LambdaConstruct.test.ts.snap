// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LambdaConstruct synthetizes properly 1`] = `
"{
  \\"Resources\\": {
    \\"LambdaServiceRoleFC57F673\\": {
      \\"Type\\": \\"AWS::IAM::Role\\",
      \\"Properties\\": {
        \\"AssumeRolePolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"sts:AssumeRole\\",
              \\"Effect\\": \\"Allow\\",
              \\"Principal\\": {
                \\"Service\\": \\"lambda.amazonaws.com\\"
              }
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"ManagedPolicyArns\\": [
          {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"arn:\\",
                {
                  \\"Ref\\": \\"AWS::Partition\\"
                },
                \\":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\\"
              ]
            ]
          }
        ]
      }
    },
    \\"Lambda04A6F88D\\": {
      \\"Type\\": \\"AWS::Lambda::Function\\",
      \\"Properties\\": {
        \\"Code\\": {
          \\"S3Bucket\\": {
            \\"Ref\\": \\"AssetParametersf30060edd2b28691a1d6d001fe4d2720bccc8f276171413ea467438e29c6edfaS3BucketB44EDA8B\\"
          },
          \\"S3Key\\": {
            \\"Fn::Join\\": [
              \\"\\",
              [
                {
                  \\"Fn::Select\\": [
                    0,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParametersf30060edd2b28691a1d6d001fe4d2720bccc8f276171413ea467438e29c6edfaS3VersionKeyD9880699\\"
                        }
                      ]
                    }
                  ]
                },
                {
                  \\"Fn::Select\\": [
                    1,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParametersf30060edd2b28691a1d6d001fe4d2720bccc8f276171413ea467438e29c6edfaS3VersionKeyD9880699\\"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        \\"Role\\": {
          \\"Fn::GetAtt\\": [
            \\"LambdaServiceRoleFC57F673\\",
            \\"Arn\\"
          ]
        },
        \\"Handler\\": \\"index.handler\\",
        \\"Runtime\\": \\"nodejs14.x\\"
      },
      \\"DependsOn\\": [
        \\"LambdaServiceRoleFC57F673\\"
      ]
    }
  },
  \\"Parameters\\": {
    \\"AssetParametersf30060edd2b28691a1d6d001fe4d2720bccc8f276171413ea467438e29c6edfaS3BucketB44EDA8B\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 bucket for asset \\\\\\"f30060edd2b28691a1d6d001fe4d2720bccc8f276171413ea467438e29c6edfa\\\\\\"\\"
    },
    \\"AssetParametersf30060edd2b28691a1d6d001fe4d2720bccc8f276171413ea467438e29c6edfaS3VersionKeyD9880699\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 key for asset version \\\\\\"f30060edd2b28691a1d6d001fe4d2720bccc8f276171413ea467438e29c6edfa\\\\\\"\\"
    },
    \\"AssetParametersf30060edd2b28691a1d6d001fe4d2720bccc8f276171413ea467438e29c6edfaArtifactHashA7F6260D\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"Artifact hash for asset \\\\\\"f30060edd2b28691a1d6d001fe4d2720bccc8f276171413ea467438e29c6edfa\\\\\\"\\"
    }
  }
}"
`;
